### ISSUE TITLE
More options with Arrays

### ISSUE THREAD

**Issue by pajenkin** [yargs/yargs#167]
Hello, as mentioned in #164 (this was correctly fixed) , the way to parse args was just modified, however, for those of us importing into other systems, having spaces in the command line that is also consumed by another program(ex: gulp), it would be great to have more options for parsing on the -cli. 

For Example: 

```
#commandline
$ foo parse --test1 what is this test

#translation parsed
foo *option parse* flag--test1-['what'] *task is* *task this* *task test*
```

Because it reads it as another task seperated by a white space. 

Giving more options that would allow you to pass in params in different ways would be nice. Either ways to make a delimited list, a way to pass in the delimiter or have enclosures:

```
#commandline 
$ foo parse --test1 [what,is,this,test]
$ foo parse --test1 "what,is,this,test"
$ foo parse --test1 what,is,this,test
$ foo parse --test1 -| what|is|this|test #pass "|" as the delimiter

#translation
foo *option parse* flag--test1-['what','is','this','test']
```

This would be a 'nice to have,' but would like to point out the issue if you want to have a bit more translation between task-runners or anything that receives commandline arguments seperated by a space.

**Comment by bcoe** [comment]
@pajenkin would you be able to use the approach of passing the flag multiple times:

`$0 --arr hello --are hello2`

**Comment by pajenkin** [comment]
@bcoe Yes, I can use the approach of passing the flag multiple times, but if building something that is meant for a user who is not familiar with the system, that creates the dependency of knowing how to use yargs. I would like to think that the more intuitive way to pass an array would be to pass objects in a string with a delimiter or in brackets or separated in white-space (which is why I referenced #164, but I am not able to use white-space ). I know this is another use case, but with the rise of build systems and more items being built through command-line, it seems like a good addition. 

I know this would be an "enhancement" but it would make it much more compatible with systems that use your module (which is awesome btw).

**Comment by bcoe** [comment]
@pajenkin I want to be careful not to add too many features to yargs, it can get a bit confusing to sifting through all of the documentation.

One option to get the behavior you're looking for, would be to simply accept the array as a string argument, and then perform a post-processing step after you parse the args, so:

```js
argv.foo = argv.foo.split(',')

// or

argv.foo = JSON.parse(argv.foo);
```

One feature that comes to mind, rather than more flexible arrays, could we add a post-processing hook so that you could do this:

```js
var argv = require('yargs')
  .option('foo', {alias: 'f'})
  .post('foo', function (foo) {
     return foo.split(',');
  })
```

```shell
./my-bin --f=a,b,c,d
```

```js
console.log(argv.f, argv.foo).
```

```shell
['a', 'b', 'c', 'd']
['a', 'b', 'c', 'd']
```

I'd rather introduce a more generic post-processing step like this, than add too much more complexity to arrays. Thoughts? CC: @nylen, @getify

**Comment by pajenkin** [comment]
I like both of those suggestions but would lean toward the latter. That would really allow for flexibility and might actually allow for something like #166 to get more options for the type of return arrays. Then the user can map however they want.

**Comment by getify** [comment]
These options are all (potentially) backwards-incompatible:

```
$ foo parse --test1 [what,is,this,test]
$ foo parse --test1 "what,is,this,test"
$ foo parse --test1 what,is,this,test
```

Someone may already be passing those kinds of values and not wanting them to be assumed as separate items. The option of passing your own delimiter (to override <space>) seems most interesting to me.

I would observe that the balance is probably between authorability vs toolability. That is, `yargs` should air on the side of things that are easy for a user to enter onto the CLI themselves (hence, the default to <space>) and allow overriding for tooling options. In that sense, a tool that's automatically generating the CLI command can probably easily produce the `--flag x --flag y ..` syntax.

**Comment by MadLittleMods** [comment]
:+1: for post-processing transforms, related to https://github.com/yargs/yargs/issues/191

The current [`opts.configParser`](https://www.npmjs.com/package/yargs#optionskey-opt) is also similar to what is needed but no extra processing after handing it off.

**Comment by nexdrew** [comment]
Note that custom "post-processing" of option values is implemented as _coercion_ with #586, which will be in the upcoming 5.0.0 release. 🎉 This would allow you to e.g. expand a string value into an array.

Also note that yargs now supports variadic positional args (for a command), which would allow you to do this:

```js
require('yargs')
  .command('parse <one> [others..]', 'Parse stuff', {}, function (argv) {
    // argv.one is a single required value, argv.others is an array of optional values
    var flag = '--' + argv.one + ' '
    console.log(flag + argv.others.join(' ' + flag))
  })
  .argv
```

```console
$ node issue167.js parse test1 what is this test
--test1 what --test1 is --test1 this --test1 test
```

**Comment by bcoe** [comment]
@pajenkin @MadLittleMods @getify `yargs@5.x` is now the `next` candidate, with support for post-processing:

https://github.com/yargs/yargs/pull/586

To give it a spin run:

`npm cache clear; npm i yargs@next`

This should give you the ability to implement custom array parsing logic.
